### PR TITLE
[SPARK-43778][SQL]Reassign expression IDs in the scalar subquery input

### PR DESCRIPTION
**Pull request by agubichev**
### What changes were proposed in this pull request?

Reassign expression IDs in the scalar subquery to avoid duplicate expression IDs.
The PR introduces a new logical rule that reassigns the IDs in the given logical plan. This rule may later be used in other areas of optimizer/analyzer.

### Why are the changes needed?

Correctness bug, see https://issues.apache.org/jira/browse/SPARK-43778

### Does this PR introduce _any_ user-facing change?

Fixes a correctness issue.

### How was this patch tested?

Query test.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/AssignNewExprIds.scala [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+import scala.collection.mutable
+
+import org.apache.spark.sql.catalyst.expressions.{Alias, Attribute, Expression, ExprId, NamedExpression}
+import org.apache.spark.sql.catalyst.expressions.aggregate.{AggregateExpression, Partial, PartialMerge}
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.rules.Rule
+
+/**
+ * Reassigns expression IDs in every expression of the given LogicalPlan (including subqueries
+ * contained in the plan).
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/AssignNewExprIds.scala:29]: Maybe also mention that:
* this rule applies to CTEs as well as subquery expressions
* after this rule finishes, we guarantee that all expression IDs for all expressions in the plan are different from the original plan, and:
* all of the original expression IDs maps consistently to the same new expression ID and this mapping holds throughout the entire plan
* the purpose of this rule is to support cases like safely adding multiple copies of the same subplans to the main plan, such as for self-joins.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/AssignNewExprIds.scala [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+ */
+object AssignNewExprIds extends Rule[LogicalPlan] {
+
+  private def assignNewExprIdsinExpr(input: Expression,
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/AssignNewExprIds.scala:33]: Scala lets you put the method definitions in any order :) maybe put the `apply` first, that's the convention for Catalyst rules, and then the rest of the methods following in the order they're referenced from `apply`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/AssignNewExprIds.scala [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+                                     reassignedExprs: mutable.HashMap[ExprId, Attribute]):
+  Expression = input match {
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/AssignNewExprIds.scala:35]: the style guide says we should put the `: ReturnType` on the same line as the last paramaeter

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/AssignNewExprIds.scala [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    case a: Attribute =>
+      val newAttribute = reassignedExprs.getOrElse(a.exprId,
+        a.withExprId(NamedExpression.newExprId))
+      reassignedExprs.put(a.exprId, newAttribute)
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/AssignNewExprIds.scala:39]: can we avoid the `put` if the key was actually present in the map? e.g.

```
reassignedExprs.get(a.exprId).getOrElse {
  val newAttribute = a.withExprId(NamedExpression.newExprId)
  reassignedExprs.put(a.exprId, newAttribute)
  reassignedExprs.put(newAttribute.exprId, newAttribute)
  newAttribute
}
```

same for the `Alias` case below.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala [hunk 5]
```diff
@@ -32,9 +32,11 @@ import org.apache.spark.sql.catalyst.rules._
 import org.apache.spark.sql.catalyst.trees.TreePattern.{EXISTS_SUBQUERY, IN_SUBQUERY, LATERAL_JOIN, LIST_SUBQUERY, PLAN_EXPRESSION, SCALAR_SUBQUERY}
 import org.apache.spark.sql.errors.{QueryCompilationErrors, QueryExecutionErrors}
 import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.internal.SQLConf.REASSIGN_IDS_IN_SCALAR_SUBQUERY
 import org.apache.spark.sql.types._
 import org.apache.spark.util.Utils
 
+
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:39]: revert extra newline?

**Review by dtenedor — COMMENTED** [review]
It will be very useful to have this rule in Spark, thank you for adding it!!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/AssignNewExprIds.scala (hunk 4 above)
Comment (agubichev) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/AssignNewExprIds.scala:39]: neat!
I think with Alias, we always have to `put` as we always update the expression id.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/AssignNewExprIds.scala (hunk 1 above)
Comment (agubichev) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/AssignNewExprIds.scala]: done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/AssignNewExprIds.scala (hunk 2 above)
Comment (agubichev) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/AssignNewExprIds.scala:47]: done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/AssignNewExprIds.scala [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+                               reassignedExprs: mutable.HashMap[ExprId, Attribute]):
+  LogicalPlan = {
+    transformUpAllExpressions(plan, {
+      case e: Expression => assignNewExprIdsinExpr(e, reassignedExprs)
+    })
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = {
+    val reassignedExprs = mutable.HashMap.empty[ExprId, Attribute]
+    assignNewExprIds(plan, reassignedExprs)
```
[the remark this answers is not in the document]
Comment (agubichev) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/AssignNewExprIds.scala]: done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/AssignNewExprIds.scala (hunk 3 above)
Comment (agubichev) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/AssignNewExprIds.scala]: done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/AssignNewExprIds.scala [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+      reassignedExprs.put(a.exprId, newAlias.toAttribute)
+      reassignedExprs.put(newAlias.exprId, newAlias.toAttribute)
+      newAlias
+    case a: AggregateExpression =>
+      if (a.mode == PartialMerge || a.mode == Partial) {
+        // Partial aggregation's attributes are going to be reused in the final aggregations.
+        // In order to avoid renaming attributes of final aggregations, keep the intermediate
+        // attributes as is.
+        reassignedExprs.put(a.resultAttribute.exprId, a.resultAttribute)
+        return a
```
[the remark this answers is not in the document]
Comment (agubichev) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/AssignNewExprIds.scala]: done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala (hunk 5 above)
Comment (agubichev) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala]: done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala [hunk 8]
```diff
@@ -598,26 +600,43 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] with AliasHelpe
...
-          currentChild.output :+ origOutput,
-          Join(currentChild, query, LeftOuter, conditions.reduceOption(And), joinHint))
+
+        // Reassign expression IDs in the future right side of the join to avoid ID conflicts.
+        val newQuery = if (SQLConf.get.getConf(REASSIGN_IDS_IN_SCALAR_SUBQUERY)) {
+          AssignNewExprIds(query)
+        } else {
+          query
+        }
+        val newOutput = newQuery.output.head.withNullability(true)
```
[the remark this answers is not in the document]
Comment (agubichev) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala:610]: done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 9]
```diff
@@ -4280,6 +4280,13 @@ object SQLConf {
       .checkValue(_ >= 0, "The threshold of cached local relations must not be negative")
       .createWithDefault(64 * 1024 * 1024)
 
+  val REASSIGN_IDS_IN_SCALAR_SUBQUERY =
+    buildConf("spark.sql.optimizer.reassignIdsInScalarSubquery.enabled")
+      .internal()
+      .doc("Reassigns expression IDs in scalar subqueries.")
```
[the remark this answers is not in the document]
Comment (agubichev) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: done

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/AssignNewExprIdsSuite.scala [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        .where($"e" === $"a")
+    val subqExpr = ScalarSubquery(subqPlan)
+    val originalQuery =
+      testRelation1
+        .select(subqExpr.as("sum"))
+        .join(testRelation2, joinType = LeftSemi, condition = Some($"sum" === $"d")).analyze
+    check(originalQuery)
+  }
+
+  test("Plan with IN subquery") {
```
[the remark this answers is not in the document]
Comment (agubichev) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/AssignNewExprIdsSuite.scala:74]: done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/AssignNewExprIds.scala [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+        newAttribute
+      }
+    case a: Alias =>
+      val newAlias = Alias(a.child, a.name)(NamedExpression.newExprId,
+        a.qualifier, a.explicitMetadata, a.nonInheritableMetadataKeys)
+      reassignedExprs.put(a.exprId, newAlias.toAttribute)
+      reassignedExprs.put(newAlias.exprId, newAlias.toAttribute)
+      newAlias
+    case a: AggregateExpression =>
+      if (a.mode == PartialMerge || a.mode == Partial) {
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/AssignNewExprIds.scala:64]: As discussed offline, let's leave this part out for now, it is OK if we don't update these internal attribute reference IDs for now, in the interest of fixing this correctness bug in a simple way.

**Pull request closed without merging**